add internal librlist library and flux-R utility for working with Rv1

grondo opened the pull request:
This PR moves the sched-simple `rlist` code (a simple implementation representing a list of resource nodes) to `src/common/librlist` so it may be used by other flux-core components (e.g. potentially the `resource` module).

The implementation of rlist is then expanded and improved to include what I assumed were requirements for a generic resource set library. Part of the purpose here is to learn what might be required and useful for an exported libflux-rset API.

Finally, a new utlity `flux-R` (for lack of a better name) is added which uses librlist to encode, decode, and combine R objects.  E.g.:

```
$ flux R
Usage: flux-R append 
   or: flux-R decode [OPTIONS]...
   or: flux-R diff 
   or: flux-R encode [OPTIONS]...
   or: flux-R intersect 
   or: flux-R rerank 
Read, generate, and process RFC 20 Resource Set objects.
Options:
  -h, --help             Display this message.
```

`flux-R encode` can be used to quickly generate Rv1 objects for testing:
```
$ flux R encode --help
Usage: flux-R encode [OPTIONS]...

Encode RFC 20 R objects for testing.

Options:

  -h, --help             Display this message.
  -H, --hosts=HOSTS      Generate R with nodelist set to HOSTS. By default,
                         duplicate the local hostname to match the number of
                         ranks given in --ranks.
  -c, --cores=IDS        Assign cores with IDS to each rank in R. Default is to
                         assign a single core "0" to each rank.
  -g, --gpus=IDS         Assign gpu resources with IDS to each rank in R.
                         Default is to assign no gpu resources.
  -r, --ranks=IDSET      Generate an R with ranks in IDSET. If not provided then
                         either match the number of nodes given  in --hosts
                         option, or emit a single rank: "0"

$ flux R encode -r 0-3 -c 0-3 -g 0 -H fluke[0-3]
{"version": 1, "execution": {"R_lite": [{"rank": "0-3", "children": {"gpu": "0", "core": "0-3"}}], "nodelist": ["fluke[0-3]"]}}
```

To encode a set of non-homogeneous ranks, `flux R append` will append all R objects on stdin:
```
$ (flux R encode -r 0-3 -c 0-3 -g 0 -H fluke[0-3] && flux R encode -r 4-7 -c 0-3 -H fluke[4-7]) | flux R append
{"version": 1, "execution": {"R_lite": [{"rank": "0-3", "children": {"core": "0-3", "gpu": "0"}}, {"rank": "4-7", "children": {"core": "0-3"}}], "nodelist": ["fluke[0-7]"]}}
```

`flux R decode` takes the union of all R objects on stdin and emits Rv1 object by default, but offers other options to modify or print stats from the result:
```
$ flux R decode --help
Usage: flux-R decode [OPTIONS]...

Return the union of all R objects on stdin and print details or summary of the
result. By default an RFC 20 JSON object is emitted on stdout, unless one or
more options below are used

Options:

  -h, --help             Display this message.
  -c, --count=TYPE       Print count of resource TYPE in R (ranks, core, gpu)
  -i, --include=RANKS    Include only specified ranks.
  -n, --nodelist         Print nodelist in hostlist form from R, if any
  -r, --ranks            Print ranks in idset form from R, if any
  -s, --short            Print short-form representation of R
$ (flux R encode -r 0-3 -c 0-3 -g 0 -H fluke[0-3] && flux R encode -r 4-7 -c 0-3 -H fluke[4-7]) | flux R decode --short
rank[0-3]/core[0-3],gpu0 rank[4-7]/core[0-3]
```

Finally, there are frontends in `flux-R` for the rlist `intersect` and `diff` methods:
```
$ (flux R encode -r 0-3 -c 0-7 && flux R encode -r 2-7 -c 5) | flux R intersect | flux R decode --short
rank[2-3]/core5
$ (flux R encode -r 0-3 -c 0-7 && flux R encode -r 2-7 -c 5) | flux R diff | flux R decode --short
rank[0-1]/core[0-7] rank[2-3]/core[0-4,6-7]
```

Perhaps it will be useful to use these operations on a set of R from jobs, I'm not sure. For now they are largely there for testing and demonstration.

Still TODO:

 * Need @garlick's review to determine what requirements are missing for the core `resource` module.

 * The current implementation drops any `.scheduling` (and `starttime`,`expiration` for that matter) from the input R object. I expect that support for carrying along the opaque key may be relegated to a public `libflux-rset` API, which may wrap `librlist` while adding support for these other keys specified in the RFC. It is unclear what could be done with the `.scheduling` data during set operations like `intersect` and `diff` (likely have to keep the value from the first set intact, and drop the value from the second)

 * I still need to go back through the code at least once and do some hardening and potential refactoring. Unfortunately the process to get to a usable librlist did not end up straightforward as I'd hoped, so the code quality may not be quite up to par. (However, given that this will all be replaced by Rv2 at some point, it may be diminishing returns to spend *too* much time on that)

garlick commented:
Here are some results of a quick scan of the `byrank.h` functions in pr #3265 vs `rlist.h` here, since the plan is to rebase that PR on this one and dispose of `byrank.[ch]`.  I just listed the functions in `byrank.h` that don't have direct `rlist.h` counterparts and how I would compensate, glossing over the distinction between `R` and `rlist` objects.  Most of the differences are easily resolved in the resource module or with trivial additions here.

The one area that probably requires further discussion is how to validate a configured R against a local hwloc topology, see below.

`byrank_resobj_sub()` - remove excluded ranks from R returned in scheduler.acquire.
 * call `rlist_ranks()` to get idset included in R
 * remove excluded ranks from idset
 * `rlist_copy_ranks()` to get filtered R

`byrank_resobj_from_hwloc()` -  create R fragment from `hwloc_topology_t` result of `local_topology_load()`
* add XML version of `local_topology_load()`
* call `rlist_from_hwloc()` on XML result
* (alternatively) add `hwloc_topology_t` version of `rlist_from_hwloc()`

`byrank_resobj_from_xml()` - create R from set of XML strings indexed by rank in a json_t
 * foreach rank in json_t:
   - call `rlist_from_hwloc()` on `json_string_value(val)`
   - call `rlist_union()` to build single rlist

`byrank_validate()` - check "probed" R fragment against configured R
* `rlist_get_rank() to get rank's portion of R`
* `rlist_diff()` R fragment with above

Problem: need to discuss what should trigger a failure here.  I had assumed probed resources that were missing from R were OK, but diff would call that out.  Also, need some mechanism for generating a human readable error for any problems.

`byrank_check_ranks()` - detect ranks in R out of [0:size-1] range
* call idset_ranks()
* check that idset_last() < size

chu11 commented on src/common/librlist/rnode.h:
```diff
@@ -17,28 +17,74 @@
+                                     const char *name,
+                                     const char *ids);
+
+enum rnode_copy_flags {
```
perhaps add a comment here.  wasn't entirely clear what these flags did.

chu11 commented on src/common/librlist/rnode.c:
```diff
@@ -12,72 +12,436 @@
+        goto fail;
+    if (!(n->cores = zhashx_lookup (n->children, "core")))
+        goto fail;
+    if (flags == RNODE_COPY_ALLOC && rnode_set_alloc (n) <= 0)
```
to me `flags` suggests bitwise AND here not `==`.  Should `flags` be an enum type?

grondo replied:
yeah, I wanted to leave it open to flags in the future. What the function is doing under the covers should be hidden from the api user.

chu11 reviewed (commented):
did a very lighthearted skim just to familiarize myself a bit more with the code, few nits i found.

grondo commented:
> The one area that probably requires further discussion is how to validate a configured R against a local hwloc topology, see below.

Ah yeah, sorry! I meant to mention that one in the TODO.

> remove excluded ranks from R returned in scheduler.acquire.

It would also be trivial to add `rlist_exclude_ranks()` and drop ranks without all the copying. Alternately, you could use the `rlist_mark_down()` and then `rlist_copy_empty()` (which should perhaps  be renamed to `rlist_copy()`.

> byrank_resobj_from_hwloc() - create R fragment from hwloc_topology_t result of local_topology_load()

Maybe I don't understand, but this is what `rlist_from_hwloc()` does when `xml == NULL`.

garlick commented:
> Maybe I don't understand, but this is what rlist_from_hwloc() does when xml == NULL.

Sorry! Missed that.  Problem solved.

> It would also be trivial to add rlist_exclude_ranks() and drop ranks without all the copying. Alternately, you could use the rlist_mark_down() and then rlist_copy_empty() (which should perhaps be renamed to rlist_copy())

This one I'm not worried about.  Creating a list of unexcluded ranks is trivial.

One other thing that might be useful is to put a `rhwloc_` prefix on the `rhwloc.h` functions so those of us not that familiar with the hwloc API don't mistake those functions for native hwloc ones.

grondo commented:
> One other thing that might be useful is to put a rhwloc_ prefix on the rhwloc.h functions so those of us not that familiar with the hwloc API don't mistake those functions for native hwloc ones.

Those functions should not be used outside of librlist, maybe I should put them back inside of `rlist.c` where they started? It was only broken out for testing. However, renaming the functions is a good idea so I can do that too.

grondo commented:
Forced a rebase and push with the following changes:

 * Added `rlist_verify (rlist_error_t *errp, struct rlist *expected, struct rlist *actual)` to verify that the single rank rlist in `actual` has the expected resources for its matching rank in `expected`.
 * Added `rlist_remove_ranks()` to explicitly remove a set of ranks from an rlist object
 * Added `flux R decode --remove` option and `flux R verify` command
 * Add limited support for pass-through opaque `.scheduling` key
 * Add support for `starttime` and `expiration` keys
 * Removed the misnamed `flags` argument to `rnode_copy` and instead use separate `rnode_copy*` calls
 * Rename all calls in `rhwloc.h` with `rhwloc_*` prefix. Add `char * rhwloc_local_topology_xml ()` to return local topology XML string.
 * Expand unit tests and `flux-R` based sharness tests

grondo commented:
Pushed an update to add a remapping function `rlist_remap()` which will remap resource IDs from an Rv1 object in addition to ranks. Currently `rlist_remap()` ignores "gpu" resources as discussed in #3193, however this is configurable if needed via an internal hash of "noremap" resource names.

grondo commented:
Ok, I've rebased and squashed a lot of the incremental work down into just a few commits. Unfortunately there wasn't a good way to present the set of enhancements to librlist as an orderly set of commits, so there is just one big "improve librlist" commit.

I've removed WIP since I believe the work here is functional.

garlick commented on src/common/librlist/rlist.c:
```diff
@@ -79,192 +218,460 @@ struct rlist *rlist_copy_down (const struct rlist *orig)
-            n = rnode_create_idset (n->rank, n->ids);
-            if (!n || !zlistx_add_end (rl->nodes, n))
+            struct rnode *copy = rnode_copy_empty (n);
+            if (!copy || rlist_add_rnode_new (rl, copy))
```
for consistency with other uses, check for `rlist_add_rnode_new () < 0`?

garlick commented on src/common/librlist/rlist.c:
```diff
@@ -79,192 +218,460 @@ struct rlist *rlist_copy_down (const struct rlist *orig)
+    while (i != IDSET_INVALID_ID) {
+        if ((n = rlist_find_rank (rl, i))) {
+            struct rnode *copy = rnode_copy (n);
+            if (!copy || rlist_add_rnode_new (result, copy)) {
```
ibid

garlick reviewed (approved):
I've taken a quick spin through this and didn't find anything.  Well two instances of one extremely minor nit (OK to ignore).  As discussed offline, since this is an internal library shared among sched-simple and resource, we can defer for now the "hardening" against bad input parameters and errno handling.

I'm using this successfully in the resource module to manipulate and reduce Rv1 objects without ever having to crack one open, IMHO a good measure of successful interface design.  The ones I'm using are
`rlist_from_json()`
`rlist_to_R()`
`rlist_remove_ranks()`
`rlist_ranks()`
`rlist_remap()`
`rlist_verify()`
`rlist_from_hwloc()`
`rlist_copy_empty()`
`rlist_append()`
`rlist_destroy()`

Coverage is good. I'd say when you're happy, just set MWP.

grondo commented:
Added some commits (which will eventually need to be squashed) which add some necessary functionality to librlist:

 * rlist_assign_hosts() - used by flux-R encode --local -H HOSTS
 * rlist_rerank() - used to reassign ranks in R based on a hostlist (from [bootstrap])

garlick commented:
New changes LGTM!

grondo commented:
> Edit: commit validator doesn't like long subject in one of your commits though

That's ok, those were meant to be squashed.

I've now squashed the new commits and forced a push

grondo commented:
I did get a version of libhwloc compiled with cuda support via
```
$ module load cudatoolkit
$ CFLAGS="-I/opt/cudatoolkit/11.0/include" LDFLAGS="-L/opt/cudatoolkit/11.0/lib64" ./configure --enable-cuda --enable-libnuma
```

It looks like libhwloc internally does support `CUDA_VISIBLE_DEVICES`, not just lstopo, e.g.
```
ƒ(s=1,builddir) grondo@pascal7:~/git/flux-core.git$ flux R encode --local | flux R decode --short
rank0/core[0-35],gpu[0-1]
ƒ(s=1,builddir) grondo@pascal7:~/git/flux-core.git$ CUDA_VISIBLE_DEVICES=1 flux R encode --local | flux R decode --short
rank0/core[0-35],gpu0
ƒ(s=1,builddir) grondo@pascal7:~/git/flux-core.git$ CUDA_VISIBLE_DEVICES=0 flux R encode --local | flux R decode --short
rank0/core[0-35],gpu0
```

garlick commented:
Restarted builder that hit this in the valgrind test:
```
2020-10-30T03:00:31.433454Z broker.err[0]: rc1.0: flux: flux_open: No such file or directory
2020-10-30T03:00:31.478766Z broker.err[0]: rc1.0: /bin/bash -c /usr/src/etc/rc1 Exited (rc=1) 0.7s
```

garlick commented:
This one ready to go in?  Please set MWP if so.

grondo commented:
Ok, setting MWP.
